connectors/utils.py

```diff
@@ -338,3 +339,57 @@ async def _wait_for_room(self, item):
     async def put(self, item):
         item_size = await self._wait_for_room(item)
         return await super().put((item_size, item))
+
+
+class ConcurrentTasks:
+    """Async task manager.
+
+    Can be used to trigger concurrent async tasks with a maximum
+    concurrency value.
+
+    - `max_concurrency`: max concurrent tasks allowed, default: 5
+    - `results_callback`: when provided, synchronous funciton called with the result of each task.
+    """
+
+    def __init__(self, max_concurrency=5, results_callback=None):
+        self.max_concurrency = max_concurrency
+        self.tasks = []
+        self.results_callback = results_callback
+        self._task_over = asyncio.Event()
+
+    def __len__(self):
+        return len(self.tasks)
+
+    def _callback(self, task, result_callback=None):
+        self.tasks.remove(task)
+        self._task_over.set()
+        if task.exception():
+            raise task.exception()
+        if result_callback is not None:
+            result_callback(task.result())
+        # global callback
+        if self.results_callback is not None:
+            self.results_callback(task.result())
+
+    async def put(self, coroutine, result_callback=None):
+        """Adds a coroutine for immediate execution.
+
+        If the number of running tasks reach `max_concurrency`, this
+        function will block and wait for a free slot.
+
+        If provided, `result_callback` will be called when the task is done.
+        """
+        # If self.tasks has reached its max size, we wait for one task to finish
+        if len(self.tasks) >= self.max_concurrency:
+            await self._task_over.wait()
+            # rearm
+            self._task_over.clear()
+        task = asyncio.create_task(coroutine())
+        self.tasks.append(task)
+        task.add_done_callback(
+            functools.partial(self._callback, result_callback=result_callback)
+        )
+        return task
+
+    async def wait(self):
```

Comment:
Suggestion: rename this to `wait_for_all`, so that it's really explicit the difference in blocking behavior happening here vs in the `put` function (which just waits for 1 task, IIUC)

Reply:
they are not exactly the same thing =>

`wait` will wait for all tasks to be finished and matches the `asyncio.wait` primitive notation. 

`put` will wait for a coroutine to be added for execution but not for the task created for that coroutine to be finished.
(that follows the `put` vs. `put_nowait` semantics from asyncio.Queue https://docs.python.org/3/library/asyncio-queue.html#asyncio.Queue.put)

But I see what you are saying, and in the spirit of short names (see PEP 8) I think we could rename `wait` to `join` and make it more explicit

That would make the class quite similar to `asyncio.Queue` https://docs.python.org/3/library/asyncio-queue.html#asyncio.Queue.join